File: python/retoolrpc/utils/polling.py
import time
from typing import Awaitable, Callable

from retoolrpc.utils.logger import Logger
from retoolrpc.utils.types import AgentServerStatus

CONNECTION_ERROR_INITIAL_TIMEOUT_MS = 50  # 50 milliseconds
CONNECTION_ERROR_RETRY_MAX_MS = 1000 * 60 * 10  # 10 minutes


async def loop_with_backoff(
    polling_interval_ms: int,
    logger: Logger,
    callback: Callable[[], Awaitable[AgentServerStatus]],
) -> AgentServerStatus:
    delay_time_ms = CONNECTION_ERROR_INITIAL_TIMEOUT_MS
    last_loop_timestamp = time.time() * 1000  # Convert seconds to ms

    while True:
        try:
            result = await callback()

            current_timestamp = time.time() * 1000  # Convert seconds to ms
            loop_duration_ms = current_timestamp - last_loop_timestamp
            last_loop_timestamp = current_timestamp
            logger.debug(
                f"Loop time: {int(loop_duration_ms)}ms, delay time: {delay_time_ms}ms, "
                f"polling interval: {polling_interval_ms}ms"
            )

            if result != "continue":
                return result

            time.sleep(polling_interval_ms / 1000)
            delay_time_ms = max(delay_time_ms // 2, CONNECTION_ERROR_INITIAL_TIMEOUT_MS)
        except Exception as err:
            logger.error(f"Error running RPC agent: {str(err)}")
            time.sleep(delay_time_ms / 1000)
            delay_time_ms = min(delay_time_ms * 2, CONNECTION_ERROR_RETRY_MAX_MS)

File: javascript/src/utils/polling.ts
import { AgentServerStatus } from '../types'
import type { LoggerService } from './logger'

const CONNECTION_ERROR_INITIAL_TIMEOUT_MS = 50
const CONNECTION_ERROR_RETRY_MAX_MS = 1000 * 60 * 10 // 10 minutes

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    const timeout = setTimeout(resolve, ms)
    timeout.unref()
  })
}

export async function loopWithBackoff(
  pollingIntervalMs: number,
  logger: LoggerService,
  callback: () => Promise<AgentServerStatus>,
): Promise<AgentServerStatus> {
  let delayTimeMs = CONNECTION_ERROR_INITIAL_TIMEOUT_MS
  let lastLoopTimestamp = Date.now()

  while (true) {
    try {
      const result = await callback()

      const currentTimestamp = Date.now()
      const loopDurationMs = currentTimestamp - lastLoopTimestamp
      lastLoopTimestamp = currentTimestamp
      logger.debug(
        `Loop time: ${loopDurationMs}ms, delay time: ${delayTimeMs}ms, polling interval: ${pollingIntervalMs}ms`,
      )

      if (result !== 'continue') {
        return result
      }
      await sleep(pollingIntervalMs)
      delayTimeMs = Math.max(delayTimeMs / 2, CONNECTION_ERROR_INITIAL_TIMEOUT_MS)
    } catch (err: unknown) {
      logger.error('Error running RPC agent', err)
      await sleep(delayTimeMs)
      delayTimeMs = Math.min(delayTimeMs * 2, CONNECTION_ERROR_RETRY_MAX_MS)
    }
  }
}
